platform/linux-generic/odp_random_std.c

```diff
@@ -4,52 +4,88 @@
  * SPDX-License-Identifier:     BSD-3-Clause
  */
 
-#include <odp_posix_extensions.h>
-#include <stdint.h>
-#include <stdlib.h>
 #include <odp/api/byteorder.h>
 #include <odp/api/cpu.h>
 #include <odp/api/debug.h>
 #include <odp_init_internal.h>
 #include <odp_random_std_internal.h>
+#include <odp_cpu.h>
 
+#include <stdint.h>
 #include <time.h>
 
-/* Assume at least two rand bytes are available and RAND_MAX is power of two - 1 */
-ODP_STATIC_ASSERT(RAND_MAX >= UINT16_MAX, "RAND_MAX too small");
-ODP_STATIC_ASSERT((RAND_MAX & (RAND_MAX + 1ULL))  ==  0, "RAND_MAX not power of two - 1");
+/*
+ * Xorshift64*, adapted from [1], and modified to return only the high 32 bits.
+ *
+ * [1] An experimental exploration of Marsaglia's xorshift generators, scrambled
+ *     Sebastiano Vigna, July 2016.
+ *     http://vigna.di.unimi.it/ftp/papers/xorshift.pdf
+ */
+static inline uint32_t xorshift64s32(uint64_t *x)
+{
+	/* The variable x should be initialized to a nonzero seed. [1] */
+	if (!*x)
+		/*
+		 * 2^64 / phi. As far away as possible from any small integer
+		 * fractions, which the caller might be likely to use for the
+		 * next seed after 0.
+		 */
+		*x = 11400714819323198485ull;
 
-static int32_t _random_data(uint8_t *buf, uint32_t len, uint32_t *seed)
+	*x ^= *x >> 12; /* a */
+	*x ^= *x << 25; /* b */
+	*x ^= *x >> 27; /* c */
+	return (*x * 2685821657736338717ull) >> 32;
+}
+
+static int32_t _random_data(uint8_t *buf, uint32_t len, uint64_t *seed)
 {
-	union {
-		uint32_t rand_word;
-		uint8_t rand_byte[4];
-	} u;
-	uint32_t i = 0, j, k;
-
-	while (i < len) {
-		u.rand_word = rand_r(seed);
-
-		/* Use two least significant bytes */
-		j = ODP_LITTLE_ENDIAN ? 0 : 2;
-		for (k = 0; k < 2 && i < len; i++, j++, k++)
-			*buf++ = u.rand_byte[j];
+	const uint32_t ret = len;
+
+	if (!_ODP_UNALIGNED && ((uintptr_t)buf & 3)) {
+		uint32_t r = xorshift64s32(seed);
+
+		if ((uintptr_t)buf & 1) {
+			*(uint8_t *)(uintptr_t)buf = r & 0xff;
+			r >>= 8;
+			buf += 1;
+			len -= 1;
+		}
+
+		if ((uintptr_t)buf & 2) {
+			*(uint16_t *)(uintptr_t)buf = r & 0xffff;
+			buf += 2;
+			len -= 2;
```

Comment:
Len underflows here if application requests a small number of bytes (e.g. len = 1).

A test case could be added to test various small lengths (and unaligned buf pointers).